### ISSUE TITLE
Modding guide: A review of the current state and ideas for an update

### ISSUE THREAD

**Issue by remyroy** [CleverRaven/Cataclysm-DDA#14170]
Modding is a great way to introduce new contributors to the game internals. It is easier than contributing to the source code and it does not require knowledge of git or github. Unfortunately, it seems like the different modding guides are out of date or lacking in content.

## The current state

The [modding page](http://en.cataclysmdda.com/pages/16/display) on the official website links to a textual guide which is dead. It links to a forum post that has not been updated since 2014. It showcase a tool that has not received any love in 2 years.

[Our contributing guide](https://github.com/CleverRaven/Cataclysm-DDA/blob/master/CONTRIBUTING.md) has no mention of modding.

The [MODDING.txt](https://github.com/CleverRaven/Cataclysm-DDA/blob/master/doc/MODDING.txt) file in the doc directory talks about modifying the core game json data files. While quite interesting, it cannot be easily distributed without potentially causing other issues. It is not self-contained like a mod in `data/mods` is. It should probably be updated to a markdown file with some formatting.

We have a few sample mods in `doc/sample_mods`. It contains interesting examples exposing some of the features available to modding. You have to look for them to find them. Examples are a great way to teach new contributors and we need more of those.

We had [this page about modding](https://webcache.googleusercontent.com/search?q=cache:gDraxyQmtpgJ:www.wiki.cataclysmdda.com/index.php%3Ftitle%3DMod_manager+&cd=3&hl=en&ct=clnk&gl=ca) on the wiki but that is dead right now. It was a little light on details but it gave many good useful references and pointers.

We had [this page about lua](https://webcache.googleusercontent.com/search?q=cache:gCyHZC2d1zkJ:www.wiki.cataclysmdda.com/index.php%3Ftitle%3DLua+&cd=1&hl=en&ct=clnk&gl=ca) on the wiki. It did a good job of listing the exposed lua objects and methods but it was light on details about what those objects and methods do. It was generated from some process. It might interesting to check if we can reuse that process.

I might have missed some of the other documentation/articles/guides about modding. If I did, please let me know.

## Suggestions

It would be nice to have a step-by-step guide on how to create your first mod showcasing some of the possibilities. It would be nice to have a reference guide on the exposed objects and methods in lua or some documentation on how to generate it. It would be nice to have documentation on how mods are detected and how they can be distributed without having to push them back into github. While getting great mods back into the official repository is still preferable, the barrier to entry for using git/github is enough to stop some from contributing.

It would be nice to have a video tutorial for those who have an easier time watching others doing it.

**Comment by pouar** [comment]
This is pretty much why I had to learn by reading the source code. Although to be fair I had to modify the C++ source as I did things that you can't currently do with the json interface. I ended up adding some of my own. Although it's not really useful for upstream which is why I didn't create a pull request

**Comment by chaosvolt** [comment]
Interesting. I will admit I did get used to modding first, and had no trouble with it.

Thing is, I flat-out ignored the wiki. I never used the wiki or doc files when I first started modding. I looked at existing mods, and eyeballed the json files for further ideas to test. When I needed more help, I pestered people on the forum., Eventually I eased into messing around on Github and source tweaks. And...well, the source code still tends to catch fire when I touch it, but still.

That does indeed imply that existing guides are less than useless, but it also means that having any such guide tell people to look at existing mods as a starting point is a valid idea.
